### ISSUE TITLE
Add reduce macros

### ISSUE THREAD

**Issue by slott56** [google/cel-spec#143]
`[1, 2, 3].reduce(s, x, s+x, 0)` for example, would compute a sum.  An alternative is `L.reduce("_+_", 0)` which would use implied variable bindings but severly limit the features available.

Specialized reductions would be available to avoid wordy constructs using the foundational `reduce` macro.

- `L.sum()` -- implicitly `L.sum(x, x)` 
- `L.min()` -- `L.min(x, x)`
- `L.max()` -- `L.max(x, x)`
- `L.count()`  -- `L.count(x, x)` == `size(L.filter(x, x))`

This would permit introduction of `mean()`, `stdev()`, `variance()` permitting CEL the be applied to statistically-based decisions. `size(L.filter(sample, sample > mean(benchmark)+3.*stdev(benchmark))) > 1`. We can then supply an appropriate benchmark value in a binding.

**Comment by TristonianJones** [comment]
This is great use case with solid examples. 

@JimLarson and I have been wanting to introduce a more robust set of macros into CEL for reducers. I'll add a mini-design doc here and start the official language change process with the CEL governance team.

**Comment by TristonianJones** [comment]
@slott56 My only question is whether you can satisfy the following equation today without the reducers and just using the existing `filter` macro and `size` method with some extensions for `mean` and `stddev` (not sure over what range of values those functions would apply based on the example though):

```
<list>.filter(sample, sample > mean(benchmark)+3.*stdev(benchmark))).size() > 1
```

**Comment by slott56** [comment]
Yes. As stated above, `L.count(x, condition)` is precisely `size(L.filter(x, condition))`. And yes, this is a redundant approach. I think a symmetry between count and sum are helpful. It makes generalization to sum of squares somewhat easier to understand.

**Comment by slott56** [comment]
The use if `int_max` and `int_min` will be a problem when working with float, duration, or timestamp values. Perhaps `min()` and `max()` could use `<range_expr>.first()` to grab the first value in the sequence, irrespective of type. If there is no first value (because the sequence is empty) this will raise an erroneous result.

**Comment by nbryant42** [comment]
At the office I've written an almost-compliant implementation of CEL in Elixir (it does deviate from spec in a few minor areas that make it a better fit for Elixir, e.g. all integers in Elixir/Erlang are bignums, so it makes a lot of sense to keep that property)

Our users do need to do a few things like `.sum()` that could be implemented as a `reduce` with more generality. My CEL implementation supports user-defined pluggable functions and macros by simply binding anonymous functions into a context map, so I'm strongly tempted to define a `reduce`. The only thing stopping me is what if upstream CEL implements it later and it ends up using a different parameter order.

I could get away with just defining things like `sum()` but my vote goes for adding this to the common language definition
